File: src/main/java/io/excaliburfrc/robot/Robot.java
package io.excaliburfrc.robot;

import com.revrobotics.CANSparkMax;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import io.excaliburfrc.robot.commands.galsearch.GalacticSearch;
import io.excaliburfrc.robot.subsystems.LEDs;
import io.excaliburfrc.robot.subsystems.LEDs.LedMode;

/**
 * This is a sample program to demonstrate the use of state-space classes in robot simulation. This
 * robot has a flywheel, elevator, arm and differential drivetrain, and interfaces with the sim
 * GUI's {@link edu.wpi.first.wpilibj.smartdashboard.Field2d} class.
 */
public class Robot extends TimedRobot {
  public RobotContainer m_robotContainer;

  /**
   * This function is run when the robot is first started up and should be used for any
   * initialization code.
   */
  @Override
  public void robotInit() {
    LiveWindow.disableAllTelemetry();
    // Instantiate our RobotContainer. This will perform all our button bindings
    m_robotContainer = new RobotContainer();
    // register the 5ms shooter velocity PID loop
    addPeriodic(
        () -> m_robotContainer.superstructure.shooter.fastPeriodic(),
        m_robotContainer.superstructure.shooter.kTimestep,
        0.003);
  }

  @Override
  public void simulationInit() {
    new GalacticSearch.SimulationManager();
  }

  @Override
  public void simulationPeriodic() {
    // periodic robot-wide sim
  }

  @Override
  public void robotPeriodic() {
    CommandScheduler.getInstance().run();
  }

  @Override
  public void autonomousInit() {
    m_robotContainer.initSubsystemStates();
    LEDs.INSTANCE.setMode(LedMode.YELLOW);
    m_robotContainer.drivetrain.setIdleMode(CANSparkMax.IdleMode.kBrake);
    m_robotContainer.getAuto().schedule();
  }

  double stopTime = 0;

  @Override
  public void disabledInit() {
    stopTime = Timer.getFPGATimestamp();
    CommandScheduler.getInstance().cancelAll();
    m_robotContainer.isLimited.set(true);
  }

  @Override
  public void disabledPeriodic() {
    if (Timer.getFPGATimestamp() - stopTime > 1.0) {
      m_robotContainer.drivetrain.setIdleMode(CANSparkMax.IdleMode.kCoast);
    }
  }

  @Override
  public void teleopInit() {
    CommandScheduler.getInstance().cancelAll();
    m_robotContainer.initSubsystemStates();
    // init teleop
  }

  @Override
  public void testInit() {}

  @Override
  public void testPeriodic() {}

  //    final Transporter transporter = m_robotContainer.superstructure.transporter;
  //    if (DriverStation.getInstance().getStickButton(1, 1)) {
  //      transporter.activate(Transporter.Mode.SHOOT);
  //    } else transporter.stop();
  //    m_robotContainer.superstructure.shooter.start(SmartDashboard.getNumber("target_rps", 0));
  //  }
}

File: src/main/kotlin/io/excaliburfrc/robot/Robot.kt
@file:JvmName("Main")

package io.excaliburfrc.robot

import edu.wpi.first.wpilibj.DriverStation
import edu.wpi.first.wpilibj.Joystick
import edu.wpi.first.wpilibj.RobotBase
import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.CommandScheduler

fun main() {
  RobotBase.startRobot { Robot }
}

const val leftX = 0

const val leftY = 1

const val rightX = 2

const val rightY = 5

/**
 * This is a sample program to demonstrate the use of state-space classes in robot simulation. This
 * robot has a flywheel, elevator, arm and differential drivetrain, and interfaces with the sim
 * GUI's [edu.wpi.first.wpilibj.smartdashboard.Field2d] class.
 */
object Robot : TimedRobot() {
  private val joystick = Joystick(0)
  private val chooser: SendableChooser<Command> by lazy {
    with(SendableChooser<Command>()) {
      setDefaultOption(
          "Arcade Drive",
          Drivetrain.arcadeDrive({ -joystick.getRawAxis(leftY) }, { joystick.getRawAxis(rightX) }))
      addOption(
          "Tank Drive",
          Drivetrain.tankDrive({ -joystick.getRawAxis(leftY) }, { -joystick.getRawAxis(rightY) }))
      addOption(
          "Left Arcade Drive",
          Drivetrain.arcadeDrive({ -joystick.getRawAxis(rightY) }, { joystick.getRawAxis(leftX) }))

      // anything else?
      // - trigger drive?
      // - curvature/cheesy drive?

      return@with this
    }
  }

  override fun robotInit() {
    Drivetrain
    SmartDashboard.putData("Drive Type", chooser)
  }

  override fun teleopInit() {
    chooser.selected?.schedule()
        ?: DriverStation.reportError("Selected was null for some flipping reason", true)
  }

  override fun teleopPeriodic() {
    CommandScheduler.getInstance().run()
  }
}
